novelist.py

```diff
@@ -0,0 +1,309 @@
+                        extracted_genres = True
+                if extracted_genres:
+                    break
+
+        if lexile_info:
+            metadata.subjects.append(SubjectData(
+                Subject.LEXILE_SCORE, lexile_info['Lexile']
+            ))
+
+        if goodreads_info:
+            metadata.measurements.append(MeasurementData(
+                Measurement.RATING, goodreads_info['average_rating']
+            ))
+
+        # If nothing interesting comes from the API, ignore it.
+        if not (metadata.measurements or metadata.series_position or
+                metadata.series or metadata.subjects or metadata.links or
+                metadata.subtitle):
+            return None
+
+        return metadata
+
+    def get_series_information(self, metadata, series_info, book_info):
+        """Returns metadata object with series info and optimal title key"""
+
+        title_key = 'main_title'
+        if series_info:
+            metadata.series = series_info['full_title']
+            series_titles = series_info.get('series_titles')
+            if series_titles:
+                matching_series_volume = [volume for volume in series_titles
+                        if volume.get('full_title')==book_info.get('full_title')]
+                if not matching_series_volume:
+                    # If there's no full_title match, try the main_title.
+                    matching_series_volume = [volume for volume in series_titles
+                        if volume.get('main_title')==book_info.get('main_title')]
+                if len(matching_series_volume) > 1:
+                    # This probably won't happen, but if it does, it will be
+                    # difficult to debug without an error.
+                    raise ValueError("Multiple matching volumes found.")
+                series_position = matching_series_volume[0].get('volume')
+                if series_position:
+                    if series_position.endswith('.'):
+                        series_position = series_position[:-1]
+                    metadata.series_position = int(series_position)
+
+                # Sometimes all of the volumes in a series have the same
+                # main_title so using the full_title is preferred.
+                main_titles = [volume.get(title_key) for volume in series_titles]
+                if len(main_titles) > 1 and len(set(main_titles))==1:
+                    title_key = 'full_title'
+
+        return metadata, title_key
+
+
+class NoveListCoverageProvider(CoverageProvider):
+
+    def __init__(self, _db, cutoff_time=None):
+        self._db = _db
+        self.api = NoveListAPI.from_config(self._db)
+        self.output_source = DataSource.lookup(self._db, DataSource.NOVELIST)
+
+        super(NoveListCoverageProvider, self).__init__(
+            "NoveList Coverage Provider", [Identifier.ISBN],
+            self.output_source, workset_size=25
+        )
+
+    def process_item(self, identifier):
```

Comment:
This needs a test. You can test it by creating a dummy API object and feeding it Metadata objects to return when lookup() is called.

If it turns out that it's a good idea to make `lookup` return a single item, as I mentioned earlier, then this should become simple enough that a test is less important. I'd probably still test it though.